kimchi/src/circuits/expr.rs

```diff
@@ -2111,5 +2135,79 @@ pub fn coeff<F>(i: usize) -> E<F> {
 
 /// You can import this module like `use kimchi::circuits::expr::prologue::*` to obtain a number of handy aliases and helpers
 pub mod prologue {
-    pub use super::{coeff, index, witness, witness_curr, witness_next, E};
+    pub use super::{coeff, constant, index, witness, witness_curr, witness_next, E};
+}
+
+#[cfg(test)]
+pub mod test {
+    use super::*;
+    use crate::circuits::{
+        constraints::ConstraintSystem, gate::CircuitGate, polynomials::generic::GenericGateSpec,
+        wires::Wire,
+    };
+    use array_init::array_init;
+    use mina_curves::pasta::fp::Fp;
+
+    #[test]
+    #[should_panic]
+    fn test_failed_linearize() {
+        // w0 * w1
+        let mut expr: E<Fp> = E::zero();
+        expr += witness_curr(0);
+        expr *= witness_curr(1);
+
+        // since none of w0 or w1 is evaluated this should panic
+        let evaluated = HashSet::new();
+        expr.linearize(evaluated).unwrap();
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_degree_tracking() {
+        // CompleteAdd(x)^7 should go over the d8 domain
+        let mut expr: E<Fp> = E::zero();
+        expr += index(GateType::CompleteAdd);
+        let expr = expr.pow(9);
```

Comment:
that's the power that pushes thing over to the next domain:

* the selector has degree n-1 (so can be tracked with n evaluations in the domain d1 of size n)
* raising a polynomial of degree n-1 to the power 7 makes it degree 7*(n-1) (and so it needs `7(n-1) + 1` evaluations)
* since `d8` is of size `8n` we are still good with that many evaluations

it'd be cool to alter the test to get exactly to a degree `8n` to see if that triggers the panic as well.

Reply:
I guess you want to update your comment then?

Reply:
ah yeah! good catch

Reply:
done